Tls auth

martinpfeifer opened the pull request:
I've never used rspec before so i just copied a bunch of stuff and changed it to fit. I don't know if it tests anything sensible :D Could you please have a look at it?

davissp14 commented:
Thanks for the PR!  I'll do my best to look through these changes in the morning.

davissp14 commented on lib/etcdv3.rb:
```diff
@@ -45,6 +45,23 @@ def token
+    # The order of the elements in the array is important
+    [cacert, key, cert].compact
+  end
+
```
This seems to work pretty well, my only nit is the loose nature of the tls_creds method.  If I am doing just server authentication, I could pass my `cacert` in as `cert, or `key` and it would still work.  There may be value in being a bit more strict here.  

Also what are your thoughts on throwing an exception when `key` is present without `cert` and vice versa?

martinpfeifer commented:
You're right. Changed it to complain with an exception if the parameters aren't passed in as they should be. Also, moved tls_creds to the private section of the class as it probably shouldn't be in the public interface.

davissp14 commented on lib/etcdv3.rb:
```diff
@@ -45,6 +45,18 @@ def token
+  def cacert
+    File.read(File.expand_path(@options.fetch(:cacert)))
+  end
+
```
I think we may want to return `nil` in the event someone hits this method when the option isn't present.  
```
irb(main):017:0* conn.key
KeyError: key not found: :key
	from /Users/shaun/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/etcdv3-0.6.0/lib/etcdv3.rb:49:in `fetch'
	from /Users/shaun/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/etcdv3-0.6.0/lib/etcdv3.rb:49:in `key'
	from (irb):17
	from /Users/shaun/.rbenv/versions/2.2.4/bin/irb:11:in `<main>'
```

Maybe something like this? 
```ruby
File.read(File.expand_path(@options[:cacert])) if @options[:cacert]
```

martinpfeifer replied:
Hmm, wouldn't it make more sense to move all those attribute readers to the private section altogether. They're only needed by the class itself and would otherwise give access to internal state which probably shouldn't be accessed by anybody directly anyway. Or am i overlooking something here?

davissp14 replied:
Yeah, that makes perfect sense to me.

martinpfeifer commented on spec/helpers/connections.rb:
```diff
@@ -9,6 +9,22 @@ def local_connection
+    end
+
+    def local_connection_with_tls_client_auth(cacert, key, cert)
+      Etcdv3.new(url: "http://#{local_url}", cacert: cacert, key: key, cert: cert)
```
Just noticed that this should be https not http, but then all tests fail because the etcd test instance is not started with any tls parameters. I'm trying to geht this to work but that can take a while. i don't really grasp rspec and all this spec helper stuff yet...

davissp14 replied:
Yeah, I am not sure off hand how to properly test this... We could certainly generate self-signed certs to use, but I think the tricky part is spinning up the Etcd with and without the tls parameters.

martinpfeifer commented:
Had to rewrite the teste quite a bit. Could you please have a look at it.

davissp14 commented:
I'm not crazy about all the stop / starting going on.  

1. We could limit the number of stop / starts by isolating the TLS specs and spin up a separate Etcd instance under a new port?  My example is a little hacky though. 

<img width="643" alt="screen shot 2017-09-04 at 8 44 07 am" src="https://user-images.githubusercontent.com/423038/30036020-2a87539c-9175-11e7-8213-1e5e7cdb9c3e.png">

2. We could assign the test_instance to a global variable.  This would allow us to stop the main Etcd instance and restart it using the TLS config.

<img width="453" alt="screen shot 2017-09-04 at 1 34 48 pm" src="https://user-images.githubusercontent.com/423038/30036116-e22ff788-9175-11e7-9fb0-a230156864f8.png">

3. We could Mock it.  


Thoughts?

davissp14 commented:
Is this still being worked on?  Would you like me to wrap up the rest?

martinpfeifer commented:
Yeah i do. Had very little time the past weeks, sorry. I'll try to push something by the end of the week.

martinpfeifer commented:
I went with starting two instances. One for plain and one for tsl. Could you please have a look at it.

davissp14 commented:
Nice, I think that looks great!  There's been ... a few changes since you last pushed.  Let me know if you need any help resolving the merge conflicts!

davissp14 commented:
Just a heads up.  I am going to finish this up if that's ok with you.

martinpfeifer commented:
sure, go ahead

davissp14 commented:
Stale.
